openhands/controller/agent_controller.py

```diff
@@ -449,7 +449,21 @@ async def _step(self) -> None:
                 == ActionConfirmationStatus.AWAITING_CONFIRMATION
             ):
                 await self.set_agent_state_to(AgentState.AWAITING_USER_CONFIRMATION)
-            self.event_stream.add_event(action, EventSource.AGENT)
+
+            if isinstance(action, CmdRunAction) and '\n' in action.command:
+                # Split the command into multiple CmdRunAction instances
+                commands = action.command.split('\n')
+                for i, cmd in enumerate(commands):
+                    if not cmd:
+                        continue
+                    new_action = CmdRunAction(command=cmd)
+                    if i < len(commands) - 1:
+                        new_action.thought = ''
+                    else:
+                        new_action.thought = action.thought
+                    self.event_stream.add_event(new_action, EventSource.AGENT)
```

Comment:
This is an interesting idea, how will it work with longer running commands though? Here `_pending_action` is set. When the first new action is executed the runtime will put an obs in the stream which will make it unset [here](https://github.com/All-Hands-AI/OpenHands/pull/4462/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eR238)... this places them in the stream and continues looping [here](https://github.com/All-Hands-AI/OpenHands/pull/4462/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eL155). Then the step would pass [here](https://github.com/All-Hands-AI/OpenHands/pull/4462/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eL394) before all are executed, no? What prevents it?

Reply:
Yeah this probably messes with _pending_action quite a bit....

@tofarr you can probably test this by asking the agent something like:
> Please run `echo hello` and `ls -lah` in the same `<execute_bash>` block